src/storage-client/src/types/connections.rs

```diff
@@ -1029,3 +1056,200 @@ impl SshTunnel {
+        self.error.lock().expect("poisoned").clone()
+    }
+}
+
+impl<C> ClientContext for DynamicBrokerRewritingClientContext<C>
+where
+    C: ClientContext,
+{
+    const ENABLE_REFRESH_OAUTH_TOKEN: bool = C::ENABLE_REFRESH_OAUTH_TOKEN;
+
+    fn rewrite_broker_addr(&self, addr: BrokerAddr) -> BrokerAddr {
+        if self.inner.overridden(&addr) {
+            return self.inner.rewrite_broker_addr(addr);
+        }
+
+        match &self.default_override {
+            Some(default_tunnel) => {
+                let broker = addr.clone();
+                let default_tunnel = default_tunnel.clone();
+                let sr = Arc::clone(&self.secrets_reader);
+                let runtime = self.runtime.clone();
+                // We spawn an additional thread here, and immediately
+                // join it so that the rdkafka thread calling this callback
+                // spends all its time parked, instead of being
+                // continuously woken up when we `block_on` the
+                // ssh tunnel creation.
+                let res = match std::thread::spawn(move || {
+                    runtime.block_on(async move {
+                        let ret = default_tunnel
+                            .build_ssh_tunnel_for_url(
+                                &*sr,
+                                &broker.host,
+                                broker.port.parse().unwrap(),
+                                "kafka default",
+                            )
+                            .await;
+                        ret
+                    })
+                })
+                .join()
+                {
+                    Ok(res) => res,
+                    Err(e) => std::panic::resume_unwind(e),
+                };
+
+                match res {
+                    Ok((local_port, token)) => {
+                        // This could do nothing if we are racing, and another
+                        // thread has overridden the broker already. Instead of
+                        // using some complicated locking scheme to avoid this,
+                        // we just accept that we may do duplicate work setting
+                        // up an extra tunnel above.
+                        self.inner.add_broker_rewrite_with_token(
+                            &format!("{}:{}", addr.host, addr.port),
+                            "127.0.0.1",
+                            Some(local_port),
+                            token,
+                        );
+
+                        let ret = self.inner.rewrite_broker_addr(addr);
+                        ret
+                    }
+                    Err(e) => {
+                        // Record the error...
+                        *self.error.lock().expect("poisoned") = Some(e.to_string());
+                        // ...and also ensure a log line shows up.
+                        self.error(KafkaError::ClientCreation(e.to_string()), &e.to_string());
```

Comment:
Maybe just `warn!` here? It's not _really_ a `KafkaError::ClientCreation`, and also I don't think we want this hitting Sentry.

Reply:
This is the way that the error is actually reported from sinks :( I don't think there is a way to avoid locking a mutex in the sink operator unless we do this...

Reply:
Ohhh, I see, it's the error collections! That makes sense. I guess in a perfect world we'd have an error variant like `KafkaError::Other(anyhow::Error)`, but probably a bit much to put that in rdkafka. Forget I said anything.